src/librslapd/src/cache.rs

```diff
@@ -78,17 +238,29 @@ pub extern "C" fn cache_char_read_begin(cache: *mut ARCacheChar) -> *mut ARCache
         debug_assert!(!cache.is_null());
         &(*cache) as &ARCacheChar
     };
+
+    cache_ref.active_read_txns.fetch_add(1, Ordering::AcqRel);
+
     let read_txn = Box::new(ARCacheCharRead {
-        inner: cache_ref.inner.read(),
+        inner: cache_ref.inner.read_stats(ReadCountStat::default()),
+        cache: cache_ref,
     });
     Box::into_raw(read_txn)
 }
 
 #[no_mangle]
 pub extern "C" fn cache_char_read_complete(read_txn: *mut ARCacheCharRead) {
     debug_assert!(!read_txn.is_null());
+
     unsafe {
-        let _drop = Box::from_raw(read_txn);
+        let read_txn_box = Box::from_raw(read_txn);
+        read_txn_box.cache.inner.try_quiesce();
```

Comment:
you need to do try_quiesce_stats(); and you need to do it *after* you call .finish.